lib/xapi-stdext-date/date.ml

```diff
@@ -44,50 +49,48 @@ type iso8601 = Ptime.t
   match x |> Ptime.of_rfc3339 |> Ptime.rfc3339_error_to_msg with
   | Error (`Msg e) -> invalid_arg (Printf.sprintf "date.ml:of_string: %s" e)
   | Ok (t, tz, _)  -> match tz with
-                      | None | Some 0 -> t
+                      | None | Some 0 -> Ptime.to_date_time t |> of_dt PrintUTC
                       | Some _        -> invalid_arg (Printf.sprintf "date.ml:of_string: %s" x)
 
-let to_string t =
-  Ptime.to_rfc3339 ~tz_offset_s:0 (* to ensure Z printed, rather than +00:00 *) t |>
-  Astring.String.filter (fun char -> char <> '-') (* remove dashes for backwards compatibility *)
+let to_string ((y,mon,d), ((h,min,s), _), print_type) =
+  match print_type with
+  | PrintUTC   -> Printf.sprintf "%04i%02i%02iT%02i:%02i:%02iZ" y mon d h min s
+  | PrintLocal -> Printf.sprintf "%04i%02i%02iT%02i:%02i:%02i" y mon d h min s
 
-let of_float x =
-  let time = Unix.gmtime x in
-  Printf.sprintf "%04d-%02d-%02dT%02d:%02d:%02dZ"
-    (time.Unix.tm_year+1900)
-    (time.Unix.tm_mon+1)
-    time.Unix.tm_mday
-    time.Unix.tm_hour
-    time.Unix.tm_min
-    time.Unix.tm_sec |> of_string
-
-(* Convert tm in localtime to calendar time, x *)
-let to_float_localtime x =
-  let datetime_to_float y mon d h min s =
-    fst Unix.(mktime { tm_year = y - 1900;
-                       tm_mon = mon - 1;
-                       tm_mday = d;
-                       tm_hour = h;
-                       tm_min = min;
-                       tm_sec = s;
-                       (* These are ignored: *)
-                       tm_wday = 0; tm_yday = 0; tm_isdst = true;
-                     })
-  in
-  let ((y, mon, d), ((h, min, s), _)) = Ptime.to_date_time x in
-  datetime_to_float y mon d h min s
+let to_ptime_t t =
+  match to_dt t |> Ptime.of_date_time with
+  | Some t -> t
+  | None ->
+    let (_, (_, offset), _) = t in
+    invalid_arg (Printf.sprintf "date.ml:to_t: dt='%s', offset='%i' is invalid" (to_string t) offset)
+
+let of_float s =
+  match Ptime.of_float_s s with
+  | None -> invalid_arg (Printf.sprintf "date.ml:of_float: %f" s)
+  | Some t -> Ptime.to_date_time t |> of_dt PrintUTC
 
 (* Convert tm in UTC back into calendar time x (using offset between above
    UTC and localtime fns to determine offset between UTC and localtime, then
    correcting for this)
 *)
-let to_float x =
-  let t = Unix.time () in
-  let offset = (t |> of_float |> to_float_localtime) -. t in
-  to_float_localtime x -. offset
+let to_float t =
+  let (_, _, print_type) = t in
+  match print_type with
+  | PrintLocal -> invalid_arg "date.ml:to_float: expected utc"
+  | PrintUTC   -> to_ptime_t t |> Ptime.to_float_s
+
+let _localtime current_tz_offset t =
+  let tz_offset_s = current_tz_offset |> Option.value ~default:0 in
+  t |> Ptime.to_date_time ~tz_offset_s |> of_dt PrintLocal
```

Comment:
Perhaps we can assert here that the offset in the final `t` is the same as `tz_offset_s` (in case Ptime rejected our offset for whatever reason)

Reply:
I've added this check:

```
let _localtime current_tz_offset t =
  let tz_offset_s = current_tz_offset |> Option.value ~default:0 in
  let localtime = t |> Ptime.to_date_time ~tz_offset_s |> of_dt PrintLocal in
  let (_, (_, localtime_offset), _) = localtime in
  if localtime_offset <> tz_offset_s then
    invalid_arg (
      Printf.sprintf "date.ml:_localtime: offsets don't match. offset='%i', t='%s'"
        tz_offset_s
        (Ptime.to_rfc3339 t)
    );
  localtime
```